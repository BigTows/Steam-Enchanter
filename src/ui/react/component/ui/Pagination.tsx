import React, { ReactNode } from "react";

interface PaginationProperties {
  items: number,
  itemsPerPage: number,
  currentPage: number;
  callback: (newPage: number) => void;
}

/**
 * (1) 2 3 4 5 6 7
 * 1 2 3 (4) 5 6 ... 8
 * (1) 2 3 ... 8
 * 1 ... 3 4 (5) 6 7 ... 44
 * 1 2 3 (4) 5 6 ... 44
 * 1 ... 39 40 (41) 42 43 44
 *
 * Like steam :)
 * @param props
 * @constructor
 */
export default function Pagination(props: PaginationProperties) {

  const maxPages = Math.ceil(props.items / props.itemsPerPage);
  const pages: number[] = [];

  if (maxPages <= 7) {
    pages.push(...Array.from({ length: maxPages }, (_, i) => i + 1));
  } else {
    for (let closePage = props.currentPage - 2; closePage <= props.currentPage + 2; closePage++) {
      if (closePage < 1 || closePage > maxPages) {
        continue;
      }
      pages.push(closePage);
    }
    if (pages[0] !== 1) {
      pages.unshift(1);
    }

    if (pages[pages.length - 1] !== maxPages) {
      pages.push(maxPages);
    }
  }

  const onPagination = (targetPage: number) => {
    if (targetPage > 0 && targetPage <= maxPages) {
      props.callback(targetPage);
    }
  };


  const pagesElements: ReactNode[] = pages.map((page, i) => {

    if (page === props.currentPage) {
      return (
        <span key={`page-${i}`}>
          {page}
          <span className="pagination_space">&nbsp;&nbsp;</span>
        </span>
      );
    } else if (i + 1 < pages.length && page + 1 !== pages[i + 1]) {
      return (
        <span key={`page-${i}`}>
          <a className="pagelink" onClick={() => onPagination(page)}>{page}</a>&nbsp;...&nbsp;
        </span>
      );
    } else {
      return (
        <span key={`page-${i}`}>
          <a className="pagelink" onClick={() => onPagination(page)}>{page}</a>
          <span className="pagination_space">&nbsp;&nbsp;</span>
        </span>
      );
    }
  });


  return (
    <div className="profile_paging">
      <div className="pageLinks">

        <a className={"pagebtn " + (props.currentPage === 1 ? "disabled" : "")}
           onClick={() => onPagination(props.currentPage - 1)}>&lt;</a>&nbsp;
        {pagesElements}
        <a className={"pagebtn " + (props.currentPage === maxPages ? "disabled" : "")}
           onClick={() => onPagination(props.currentPage + 1)}>&gt;</a>
      </div>
      Showing {((props.currentPage - 1) * props.itemsPerPage + 1).toLocaleString()}-{(Math.min(props.currentPage * props.itemsPerPage, props.items)).toLocaleString()} of {props.items.toLocaleString()} uncompleted badges
    </div>
  );
}