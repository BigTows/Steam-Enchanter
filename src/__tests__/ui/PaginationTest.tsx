import React from "react";
import renderer from "react-test-renderer";
import Pagination from "../../ui/react/component/ui/Pagination";

test("pagination with difference setup", () => {

  const callback = () => {

  };

  let component = renderer.create(
    <Pagination items={55} itemsPerPage={10} currentPage={1} callback={callback} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();


  component = renderer.create(
    <Pagination items={10} itemsPerPage={10} currentPage={1} callback={callback} />
  );
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();


  component = renderer.create(
    <Pagination items={9000} itemsPerPage={44} currentPage={15} callback={callback} />
  );
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component = renderer.create(
    <Pagination items={100} itemsPerPage={10} currentPage={10} callback={callback} />
  );
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component = renderer.create(
    <Pagination items={100} itemsPerPage={10} currentPage={9} callback={callback} />
  );
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();


  component = renderer.create(
    <Pagination items={900909} itemsPerPage={10} currentPage={4} callback={callback} />
  );
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});