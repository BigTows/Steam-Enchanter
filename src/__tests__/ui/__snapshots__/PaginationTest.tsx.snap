// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination with difference setup 1`] = `
<div
  className="profile_paging"
>
  <div
    className="pageLinks"
  >
    <a
      className="pagebtn disabled"
      onClick={[Function]}
    >
      &lt;
    </a>
     
    <span>
      1
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        2
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        3
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        4
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        5
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        6
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <a
      className="pagebtn "
      onClick={[Function]}
    >
      &gt;
    </a>
  </div>
  Showing 
  1
  -
  10
   of 
  55
   uncompleted badges
</div>
`;

exports[`pagination with difference setup 2`] = `
<div
  className="profile_paging"
>
  <div
    className="pageLinks"
  >
    <a
      className="pagebtn disabled"
      onClick={[Function]}
    >
      &lt;
    </a>
     
    <span>
      1
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <a
      className="pagebtn disabled"
      onClick={[Function]}
    >
      &gt;
    </a>
  </div>
  Showing 
  1
  -
  10
   of 
  10
   uncompleted badges
</div>
`;

exports[`pagination with difference setup 3`] = `
<div
  className="profile_paging"
>
  <div
    className="pageLinks"
  >
    <a
      className="pagebtn "
      onClick={[Function]}
    >
      &lt;
    </a>
     
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        1
      </a>
       ... 
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        13
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        14
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      15
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        16
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        17
      </a>
       ... 
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        205
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <a
      className="pagebtn "
      onClick={[Function]}
    >
      &gt;
    </a>
  </div>
  Showing 
  617
  -
  660
   of 
  9,000
   uncompleted badges
</div>
`;

exports[`pagination with difference setup 4`] = `
<div
  className="profile_paging"
>
  <div
    className="pageLinks"
  >
    <a
      className="pagebtn "
      onClick={[Function]}
    >
      &lt;
    </a>
     
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        1
      </a>
       ... 
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        8
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        9
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      10
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <a
      className="pagebtn disabled"
      onClick={[Function]}
    >
      &gt;
    </a>
  </div>
  Showing 
  91
  -
  100
   of 
  100
   uncompleted badges
</div>
`;

exports[`pagination with difference setup 5`] = `
<div
  className="profile_paging"
>
  <div
    className="pageLinks"
  >
    <a
      className="pagebtn "
      onClick={[Function]}
    >
      &lt;
    </a>
     
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        1
      </a>
       ... 
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        7
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        8
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      9
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        10
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <a
      className="pagebtn "
      onClick={[Function]}
    >
      &gt;
    </a>
  </div>
  Showing 
  81
  -
  90
   of 
  100
   uncompleted badges
</div>
`;

exports[`pagination with difference setup 6`] = `
<div
  className="profile_paging"
>
  <div
    className="pageLinks"
  >
    <a
      className="pagebtn "
      onClick={[Function]}
    >
      &lt;
    </a>
     
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        1
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        2
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        3
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      4
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        5
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        6
      </a>
       ... 
    </span>
    <span>
      <a
        className="pagelink"
        onClick={[Function]}
      >
        90091
      </a>
      <span
        className="pagination_space"
      >
          
      </span>
    </span>
    <a
      className="pagebtn "
      onClick={[Function]}
    >
      &gt;
    </a>
  </div>
  Showing 
  31
  -
  40
   of 
  900,909
   uncompleted badges
</div>
`;
